### PR TITLE
UTY-541 Add event delegates into ReaderWriter Codegen

### PR DESCRIPTION
**Pull request by mattyoung-improbable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Describe your changes here.
#### Tests
How did you test these changes prior to submitting this pull request?
What automated tests are included in this PR?
#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?
#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt [hunk 1]
```diff
@@ -10,10 +10,13 @@
 #>
 <#= generatedHeader #>
 
+using System;
```
Comment (firtina-improbable) [inline code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt]: Optional - Nit: add the usings as conditional per #evtDetails>0? Or use full names?

File: code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt [hunk 2]
```diff
@@ -78,11 +81,43 @@ namespace <#= qualifiedNamespace #>
...
+            }
+<# } #>
+<# foreach (var eventDetails in eventDetailsList) {
+            var delegateList = eventDetails.CamelCaseTypeName + "Delegates";
+            var eventType = eventDetails.EventName + "Event";
+#>
+
+            private readonly List<Action<<#= eventType #>>> <#= delegateList #> = new List<Action<<#= eventType #>>>();
+
+            // TODO check naming, should it be Xtriggered, or OnXEvent or OnX?
```
Comment (firtina-improbable) [inline code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt]: Please resolve this in this PR

File: code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt (hunk 1 above)
Comment (mattyoung-improbable) [reply code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt]: Done, fully qualified the names.

File: code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt (hunk 2 above)
Comment (mattyoung-improbable) [reply code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt]: Done. Gone with "OnX"

File: code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt [hunk 3]
```diff
@@ -78,11 +78,42 @@ namespace <#= qualifiedNamespace #>
...
+            }
+<# } #>
+<# foreach (var eventDetails in eventDetailsList) {
+            var delegateList = eventDetails.CamelCaseTypeName + "Delegates";
+            var eventType = eventDetails.EventName + "Event";
+#>
+
+            private readonly System.Collections.Generic.List<System.Action<<#= eventType #>>> <#= delegateList #> = new System.Collections.Generic.List<System.Action<<#= eventType #>>>();
+
+            public event System.Action<<#= eventType #>> On<#= eventDetails.EventName #>
```
Comment (MPeti01) [inline code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt]: This should also be declared in the user facing Reader interface

File: code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt [hunk 4]
```diff
@@ -42,6 +42,12 @@ namespace <#= qualifiedNamespace #>
 <# foreach (var fieldDetails in fieldDetailsList) { #>
             event Action<<#= fieldDetails.Type #>> <#= fieldDetails.PascalCaseName #>Updated;
 <# } #>
+
```
Comment (zeroZshadow) [inline code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt]: This newline isn't needed I think ?

Comment (firtina-improbable) [reply code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt]: +1

File: workers/unity/Assets/Generated/Source/improbable/gdk/tests/componentswithnofields/ComponentWithNoFieldsGameObjectComponentDispatcher.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
+// ===========
```
Comment (firtina-improbable) [inline workers/unity/Assets/Generated/Source/improbable/gdk/tests/componentswithnofields/ComponentWithNoFieldsGameObjectComponentDispatcher.cs]: Why are these files in this PR?

Comment (MPeti01) [reply workers/unity/Assets/Generated/Source/improbable/gdk/tests/componentswithnofields/ComponentWithNoFieldsGameObjectComponentDispatcher.cs]: They were recently generated from the merge from master and codegen was not run in the merge PR.

**Pull request merged**